Enable typechecking in _inductor/ir.py

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110112

I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

int3 committed:
Enable typechecking in _inductor/ir.py

I used a bunch of ignore-type comments for a few reasons:

1. https://github.com/pytorch/pytorch/issues/109963
2. `MutableBox.__getattr__` makes the IR fairly dynamic
3. Relatedly, a good deal of methods seem to be valid only under some
   initializations of the object, e.g. BaseView.realize_hint() is only
   valid if BaseView was initialized with an object argument that
   implements this method, but that doesn't seem to be true for all
   initializations of BaseView.

[ghstack-poisoned]

jansel commented on torch/_inductor/ir.py:
```diff
@@ -407,16 +408,31 @@ def get_reads(self):
             if self.get_reduction_type():
                 return extract_read_writes(
-                    self.make_loader(),
+                    self.make_loader(),  # type: ignore[attr-defined]
```
Can we fix this one (and most of the other attr-defined's by adding an abstract method to the base class?

int3 replied:
That isn't exactly semantics-preserving because of point 2 -- defining it on the base class (IRNode) means that MutableBox's `__getattr__` won't end up looking the method up on its `self.data` member. I did try defining one of these methods and tests did indeed fail. I do agree that this isn't very clean though...

int3 replied:
One option would be to do away with `MutableBox.__getattr__` altogether, and require every method to be explicitly defined instead. That would make evolving the IR less flexible, not sure if that's a concern at all

But we do probably need something like this at some point if we wish to enable `follow_imports` when typechecking `_inductor/` -- there would be way too many errors otherwise.

ezyang replied:
IMO MutableBox shouldn't have been implemented with a `__getattr__` but this is a hot take

jansel replied:
Agreed, you can just add an explicit "foward to .data" method in MutableBox.

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments for a few reasons:

1. https://github.com/pytorch/pytorch/issues/109963
2. `MutableBox.__getattr__` makes the IR fairly dynamic
3. Relatedly, a good deal of methods seem to be valid only under some
   initializations of the object, e.g. BaseView.realize_hint() is only
   valid if BaseView was initialized with an object argument that
   implements this method, but that doesn't seem to be true for all
   initializations of BaseView.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments for a few reasons:

1. https://github.com/pytorch/pytorch/issues/109963
2. `MutableBox.__getattr__` makes the IR fairly dynamic
3. Relatedly, a good deal of methods seem to be valid only under some
   initializations of the object, e.g. BaseView.realize_hint() is only
   valid if BaseView was initialized with an object argument that
   implements this method, but that doesn't seem to be true for all
   initializations of BaseView.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments for a few reasons:

1. https://github.com/pytorch/pytorch/issues/109963
2. `MutableBox.__getattr__` makes the IR fairly dynamic
3. Relatedly, a good deal of methods seem to be valid only under some
   initializations of the object, e.g. BaseView.realize_hint() is only
   valid if BaseView was initialized with an object argument that
   implements this method, but that doesn't seem to be true for all
   initializations of BaseView.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments for a few reasons:

1. https://github.com/pytorch/pytorch/issues/109963
2. `MutableBox.__getattr__` makes the IR fairly dynamic
3. Relatedly, a good deal of methods seem to be valid only under some
   initializations of the object, e.g. BaseView.realize_hint() is only
   valid if BaseView was initialized with an object argument that
   implements this method, but that doesn't seem to be true for all
   initializations of BaseView.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

Skylion007 commented on torch/_inductor/graph.py:
```diff
@@ -675,7 +678,7 @@ def output(self, target, args, kwargs):
             assert isinstance(
                 value, (TensorBox, sympy.Expr)
-            ), "Unsupported inductor graph input type: " + type(value)
+            ), "Unsupported inductor graph input type: " + str(type(value))
```
Just use an fstring here?

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -2588,20 +2650,21 @@ def get_unbacked_symbol_uses(self):
         # Inline constants and index_expressions
-        can_inline = (
+        if (
             hasattr(self.data, "make_loader")
```
This check is broken now since all IRNodes have a `make_loader` attr.

int3 replied:
oops. Surprised the tests didn't catch this...

I'll grep for more such `hasattr` calls.

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented:
@peterbell10 do you mind shepherding this to finish?

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 commented:
I figured out a way to convince mypy of the presence of abstract methods in IRNode without actually defining them. This approach should be safer and arguably cleaner.

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -1146,9 +1176,15 @@ def create(  # type: ignore[override]
+        # FIXME there's a bunch of dead code in this method, see https://github.com/pytorch/pytorch/issues/109963
+        # Most of the type-ignore comments below are due to this dead code
+
         def const(idx, val):
```
I think the callers are correct but the signature is wrong here, `idx` is shadowed by `inner_fn(idx)` for every use.
```suggestion
        def const(val):
```

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -1146,9 +1176,15 @@ def create(  # type: ignore[override]
-                return ops.constant(val, dst_dtype)
+                return ops.constant(
+                    val,
+                    dst_dtype,  # type: ignore[name-defined]
```
```suggestion
                    dtype,
```

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -1172,13 +1208,13 @@ def inner_fn(idx):
                 return Pointwise.create(
                     device=device,
-                    dtype=dst_dtype,
+                    dtype=dst_dtype,  # type: ignore[name-defined]
```
```suggestion
                    dtype=dtype,
```

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -3651,7 +3695,7 @@ def should_allocate(self):
 
     def __init__(self, sym, data):
-        super().__init__(None, NoneLayout(), [data])
+        super().__init__(None, NoneLayout(), [data])  # type: ignore[arg-type]
```
Would this be solved by making this function's argument `data: Buffer`?

peterbell10 commented on torch/_inductor/ir.py:
```diff
@@ -306,6 +306,21 @@ def realize(self):
+    make_loader: Callable[[], Callable[[Any], Any]]
+    make_indexer: Callable[[], Callable[[Any], Any]]
+    mark_reuse: Callable[[List[Any]], None]
+    realize_hint: Callable[[], None]
```
Ideally I think we should have stricter typing and not use `IRNode` everywhere, but this is still much better than before so I'm happy to have a workaround for now.

peterbell10 reviewed (commented):
A few minor comments, but overall this looks good.

Skylion007 commented:
We haven't upgraded to the new | Union syntax yet, please use the older version so people on older Python versions still have working type hints in their IDEs.

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 commented on torch/_inductor/codegen/cuda/cuda_template.py:
```diff
@@ -26,13 +28,13 @@ class CUDATemplate(KernelTemplate):
         super().__init__(name)
         self.input_nodes = input_nodes
-        self.output_node = Buffer("buf_out", layout)
+        self.output_node: Layout | Buffer = Buffer("buf_out", layout)
```
Yeah I think you're right, it's always a Buffer. I changed `TensorMeta.from_irnodes` to accept an immutable `Sequence[Buffer | Layout]` instead of a List to make this typecheck.

peterbell10 commented on torch/_inductor/autotune_process.py:
```diff
@@ -337,9 +338,9 @@ class TensorMeta:
+        cls, irnodes: Union[LayoutOrBuffer, Sequence[LayoutOrBuffer]]
     ) -> Union[TensorMeta, List[TensorMeta]]:
-        if isinstance(irnodes, (tuple, list)):
+        if isinstance(irnodes, collections.abc.Sequence):
```
NIT: you can use `typing.Sequence` for the isinstance check, it's equivalent.

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

int3 commented:
@pytorchbot merge

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

int3 commented:
@pytorchbot merge

int3 committed:
Update on "Enable typechecking in _inductor/ir.py"


I used a bunch of ignore-type comments, mostly due to
https://github.com/pytorch/pytorch/issues/109963.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
@pytorchbot merge
